lib/elixir/lib/module/checker.ex

```diff
@@ -0,0 +1,189 @@
+        warn(meta, state, {:undefined_module, module, fun, arity})
+
+      not function_exported?(module, fun, arity) ->
+        exports = exports_for(module)
+        warn(meta, state, {:undefined_function, module, fun, arity, exports})
+
+      true ->
+        []
+    end
+  end
+
+  # TODO: Do not warn inside guards
+  # TODO: Properly handle protocols
+  defp warn_if_undefined?(_module, :__impl__, 1, _state), do: false
+  defp warn_if_undefined?(:erlang, :orelse, 2, _state), do: false
+  defp warn_if_undefined?(:erlang, :andalso, 2, _state), do: false
+
+  defp warn_if_undefined?(module, fun, arity, state) do
+    for(
+      {:no_warn_undefined, values} <- state.compile_opts,
+      value <- List.wrap(values),
+      value == module or value == {module, fun, arity},
+      do: :skip
+    ) == []
+  end
+
+  defp exports_for(module) do
+    try do
+      module.__info__(:macros) ++ module.__info__(:functions)
+    rescue
+      _ -> module.module_info(:exports)
+    end
+  end
+
+  defp warn(meta, %{file: file, module: module, function: {fun, arity}}, warning) do
+    {warning, {file, meta[:line], {module, fun, arity}}}
+  end
+
+  defp merge_warnings(warnings) do
+    Enum.reduce(warnings, %{}, fn {warning, location}, acc ->
+      locations = MapSet.new([location])
+      Map.update(acc, warning, locations, &MapSet.put(&1, location))
+    end)
+  end
+
+  defp sort_warnings(warnings) do
+    warnings
+    |> Enum.map(fn {warning, locations} -> {warning, Enum.sort(locations)} end)
+    |> Enum.sort()
+  end
+
+  defp emit_warnings(warnings) do
+    Enum.flat_map(warnings, fn {warning, locations} ->
+      message = format_warning(warning)
+      print_warning([message, ?\n, format_locations(locations)])
+
+      Enum.map(locations, fn {file, line, _mfa} ->
+        {file, line, message}
+      end)
+    end)
+  end
+
+  defp format_warning({:undefined_module, module, fun, arity}) do
+    [
+      Exception.format_mfa(module, fun, arity),
+      " is undefined (module ",
+      inspect(module),
+      " is not available or is yet to be defined)"
+    ]
+  end
+
+  defp format_warning({:undefined_function, module, fun, arity, exports}) do
+    [
+      Exception.format_mfa(module, fun, arity),
+      " is undefined or private",
+      UndefinedFunctionError.hint_for_loaded_module(module, fun, arity, exports)
+    ]
+  end
+
+  defp format_locations([location]) do
+    format_location(location)
+  end
+
+  defp format_locations(locations) do
+    [
+      "Found at #{length(locations)} locations:\n",
+      Enum.map(locations, &format_location/1)
+    ]
+  end
+
+  defp format_location({file, line, {module, fun, arity}}) do
+    file = Path.relative_to_cwd(file)
+    line = Integer.to_string(line)
+    mfa = Exception.format_mfa(module, fun, arity)
+
+    ["  ", file, ?:, line, ": ", mfa, ?\n]
+  end
+
+  defp print_warning(message) do
+    IO.puts(:stderr, [warning_prefix(), message])
+  end
+
+  defp warning_prefix() do
```

Comment:
We have a similar function in `elixir_errors`, please expose and use it. :)

Reply:
Should we expose the prefix function or the print function?